File: Tutorial5-3bDagger-Android/src/main/java/com/example/tutorial5_1daggerandroid/di/ApplicationModule.java
package com.example.tutorial5_1daggerandroid.di;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;

import com.example.tutorial5_1daggerandroid.MainActivity;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.android.ActivityKey;
import dagger.android.AndroidInjector;
import dagger.multibindings.IntoMap;

@Module(subcomponents = {ToastMakerSubComponent.class})
public abstract class ApplicationModule {

    private static Context context;

    public ApplicationModule(Context context) {
        this.context = context;
    }

    @Provides
    @Singleton
    static SharedPreferences provideSharedPreferences() {
        System.out.println("ApplicationModule context: " + context);
        return context.getSharedPreferences("PrefName", Context.MODE_PRIVATE);
    }

    @Binds
    @IntoMap
    @ActivityKey(MainActivity.class)
    abstract AndroidInjector.Factory<? extends Activity> bindMainActivityFactory(ToastMakerSubComponent.Builder builder);
}

File: Tutorial5-2Dagger-Android-SubComponents2SeedInstance/src/main/java/com/example/tutorial5_2dagger_android_subcomponents2/di/ApplicationModule.kt
package com.example.tutorial5_2dagger_android_subcomponents2.di

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import com.example.tutorial5_2dagger_android_subcomponents2.MainActivity
import com.example.tutorial5_2dagger_android_subcomponents2.SecondActivity
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.android.AndroidInjector
import dagger.multibindings.ClassKey
import dagger.multibindings.IntoMap
import javax.inject.Singleton

@Module(subcomponents = [ToastMakerSubComponent::class, LiveObjectSubComponent::class])
abstract class ApplicationModule {

    /*
    *  NOTE: This is same as adding void inject(MainActivity mainActivity) to SubComponent as in
    *  non dagger-android, and can be removed using
    *  @ContributesAndroidInjector if SubComponent and it's Builder has NO methods
    *
    * !! In kotlin Generic should match with @ClassKey, if it's ? then Kotlin should be *
    */
// 🔥 Injects ToastMakerSubComponent objects to MainActivity
    @Binds
    @IntoMap
    @ClassKey(MainActivity::class)
    abstract fun bindMainActivityFactory(builder: ToastMakerSubComponent.Builder): AndroidInjector.Factory<*>

    // 🔥 Injects LiveObjectSubComponent objects to SecondActivity
    @Binds
    @IntoMap
    @ClassKey(SecondActivity::class)
    abstract fun bindSecondActivityFactory(builder: LiveObjectSubComponent.Builder): AndroidInjector.Factory<*>

    companion object {

        @Provides
        @Singleton
        fun provideSharedPreferences(application: Application): SharedPreferences {
            println("ApplicationModule context: $application")
            return application.getSharedPreferences("PrefName", Context.MODE_PRIVATE)
        }
    }
}